src/reversediffx.jl

```diff
@@ -0,0 +1,582 @@
+    input_derivs = deriv.(Base.tail(input))
+    P = instruction.cache[1]
+    jtvs = P(deriv(output))
+    for i in 1:length(jtvs)
+        input_derivs[i] .+= jtvs[i]
+    end
+    unseed!(output)
+    return nothing
+end
+
+@noinline function ReverseDiff.special_forward_exec!(instruction::SpecialInstruction{typeof(cat)})
+    output, input = instruction.output, instruction.input
+    dims = input[1]
+    Xs = value.(Base.tail(input))
+    out_value = cat(Xs..., dims = dims)
+    value!(output, out_value)
+    return nothing
+end
+
+################
+# Broadcasting #
+################
+
+"""
+    NotTracked(f::Function)
+
+A struct that can be used to wrap around closures, structs and arrays of structs declaring that they do not contain tracked variables. This enables a more efficient broadcasting of such functions and structs when doing automatic differentiation with `ReverseDiff` producing a `TrackedArray` instead of an `Array{<:TrackedReal}`.
+"""
+struct NotTracked{F} <: Function
+    f::F
+end
+(f::NotTracked{<:Union{Function, Type}})(args...; kwargs...) = f.f(args...; kwargs...)
+
+istypeorclosure(::F) where {F} = _istypeorclosure(F)
+istypeorclosure(::AbstractArray{F}) where {F} = _istypeorclosure(F)
+istypeorclosure(::Base.RefValue{F}) where {F} = _istypeorclosure(F)
+istypeorclosure(::AbstractArray{<:Real}) = false
+istypeorclosure(::Real) = false
+@generated _istypeorclosure(::Type{F}) where {F} = :($(fieldcount(F) > 0))
+
+@inline mayhavetracked(b) = istypeorclosure(b)
+@inline mayhavetracked(b::NotTracked) = false
+@inline mayhavetracked(b::Base.RefValue{<:NotTracked}) = false
+@inline mayhavetracked(b::Broadcasted) = mayhavetracked(b.f) || any(mayhavetracked, b.args)
+
+struct TrackedStyle <: BroadcastStyle end
+
+Broadcast.BroadcastStyle(::Type{<:Union{RTA, RTR}}) = TrackedStyle()
+Broadcast.BroadcastStyle(::TrackedStyle, b::BroadcastStyle) = TrackedStyle()
+
+# We have to re-build the original broadcast struct to get the appropriate array
+# style. We need this primarily to support CuArrays' broadcasting fixes.
+broadcast_rebuild(xs) = value(xs)
+function broadcast_rebuild(bc::Broadcasted)
+    broadcasted(bc.f, broadcast_rebuild.(bc.args)...)
+end
+
+getstyle(::Broadcasted{Style}) where {Style} = Style
+remove_not_tracked(f) = f
+remove_not_tracked(f::NotTracked) = f.f
+remove_not_tracked(f::Base.RefValue{<:NotTracked}) = Ref(remove_not_tracked(f[]))
+remove_not_tracked(f::Base.RefValue{<:NotTracked{<:AbstractArray}}) = remove_not_tracked(f[])
+function remove_not_tracked(b::Broadcasted{style}) where {style}
+    return Broadcasted{style}(remove_not_tracked(b.f), remove_not_tracked.(b.args), b.axes)
+end
+
+@generated function onlyrealarrays(args::T) where {T <: Tuple}
+    o = all(map(x -> (x <: AbstractArray{<:Real} || !(x <: AbstractArray)), T.types))
+    return :($o)
+end
+@generated function anyreals(args::T) where {T <: Tuple}
+    o = any(map(x -> x <: Real, T.types))
+    return :($o)
+end
```

Comment:
Maybe also just use regular functions here?
```julia
anyreals(args::Tuple) = tuple_type_anyreals(typeof(args))

tuple_type_anyreals(::Type{Tuple{}}) = false
function tuple_type_anyreals(::Type{T}) where T<:Tuple
	return Base.tuple_type_head(T) <: Real || tuple_type_anyreals(Base.tuple_type_tail(T))
end
```
With Julia >= 1.1, one could also use
```julia
anyreals(args::Tuple) = any(x <: Real for x in fieldtypes(typeof(args)))
```
but the former seems to be easier for the compiler.

Reply:
Fair enough, I used a generated function because it is a sure fire way of doing this at compile time. Sometimes large tuples and recursion cause inference issues but not this time it seems.

Reply:
Done.